File: modules/call-blocker/android/src/main/java/expo/modules/avoloblockers/AvoloBlockersModule.kt
package expo.modules.avoloblockers

import android.app.role.RoleManager
import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class CallBlockerModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("CallBlocker")

        Function("saveBlocklist") { prefixes: List<String>, exact: List<String> ->
            val ctx = appContext.reactContext ?: return@Function
            ctx.getSharedPreferences("blocklist", Context.MODE_PRIVATE)
                .edit()
                .putStringSet("prefixes", prefixes.toSet())
                .putStringSet("exact", exact.toSet())
                .apply()
        }

        Function("getBlocklist") {
            val ctx = appContext.reactContext
            val prefs = ctx?.getSharedPreferences("blocklist", Context.MODE_PRIVATE)
            mapOf(
                "prefixes" to (prefs?.getStringSet("prefixes", emptySet())?.toList() ?: emptyList()),
                "exact" to (prefs?.getStringSet("exact", emptySet())?.toList() ?: emptyList())
            )
        }

        Function("isRoleHeld") {
            val ctx = appContext.reactContext ?: return@Function false
            val rm = ctx.getSystemService(Context.ROLE_SERVICE) as RoleManager
            rm.isRoleAvailable(RoleManager.ROLE_CALL_SCREENING) &&
                    rm.isRoleHeld(RoleManager.ROLE_CALL_SCREENING)
        }

        Function("requestRole") {
            val activity = appContext.currentActivity ?: return@Function false
            val rm = activity.getSystemService(Context.ROLE_SERVICE) as RoleManager
            if (rm.isRoleAvailable(RoleManager.ROLE_CALL_SCREENING) &&
                !rm.isRoleHeld(RoleManager.ROLE_CALL_SCREENING)
            ) {
                val intent = rm.createRequestRoleIntent(RoleManager.ROLE_CALL_SCREENING)
                activity.startActivityForResult(intent, 1001)
                return@Function true
            }
            return@Function false
        }
    }
}

File: modules/call-blocker/ios/AvoloBlockersModule.swift
import ExpoModulesCore

public class CallBlockerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CallBlocker")

    // Call screening is Android-only; these stubs keep JS callers safe on iOS.

    Function("saveBlocklist") { (_: [String], _: [String]) in }

    Function("getBlocklist") {
      return ["prefixes": [], "exact": []] as [String: [String]]
    }

    Function("getBlockedCalls") {
      return [] as [[String: Any]]
    }

    Function("clearBlockedCalls") {
      return false
    }

    Function("isRoleHeld") {
      return false
    }

    Function("requestRole") {
      return false
    }
  }
}
